### PR TITLE
Fix a panic in cpuid/brand_string.rs

### LINKED ISSUE #2914: [Bug] Zero-length CPUID (on Chrome OS) causes panic: 'attempt to subtract with overflow'
# Describe the bug

The offending code is this:
```
        let mut len = Self::MAX_LEN;
        {
            let this_bytes = this.as_bytes();
            while this_bytes[len - 1] == 0 && len > 0 {
                len -= 1;
            }
```

You can get to a case where
```
len is 1
len -=1 -> len becomes 0
while this_bytes[unsigned 0 - 1]
```
The len check should come first to short-circuit the bad case.

## To Reproduce

On ChromeOS, download Firecracker main, go to `src/cpuid` and run `cargo test`. This gives the "attempt to subtract with overflow" error.

Alternatively, in prod, I ran through the quickstart guide and hit the much-less-helpful error
```
2022-02-28T13:38:25.634380343 [anonymous-instance:main:ERROR:src/firecracker/src/main.rs:91] Firecracker panicked at 'index out of bounds: the len is 48 but the index is 18446744073709551615', src/cpuid/src/brand_string.rs:124:19
```

## Expected behaviour

`[Author TODO: A clear and concise description of what you expected to happen.]`

## Environment

`[Author TODO: Please supply the following information):]`
`[ - Firecracker version.]`
9d8bb4c8fff7d5c082e6405f4b842979881c7a42

`[ - Architecture.]`
x86-64

`[ - Any other relevant software versions.]`
--

## Additional context

`[Author TODO: How has this bug affected you?]`
I can't even try to run firecracker

`[Author TODO: What are you trying to achieve?]`
Run firecracker

`[Author TODO: Do you have any idea of what the solution might be?]`
Yeah, I might submit a PR.

## Checks

- [x] Have you searched the Firecracker Issues database for similar problems?
- [x] Have you read the existing relevant Firecracker documentation?
- [x] Are you certain the bug being reported is a Firecracker issue?

### PR DESCRIPTION
**Pull request by wimax-grapl**
# Reason for This PR

`[Author TODO: add issue #.]`
`[Open the PR after the related issue has a clear conclusion.]`
`[If there is no issue which states the need for this PR, create one first.]`
Fixes #2914

## Description of Changes
- Fix a length check that accesses `some_slice[unsigned 0 - 1]` on systems without a Brand String
- Refactor that length check so we can write a test around it that doesn't depend on the system it's run on

- [x] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.
Yes

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: src/cpuid/src/brand_string.rs [hunk 1]
```diff
@@ -300,6 +293,21 @@ impl BrandString {
...
 }
 
+trait NullTerminatorIndex {
+    fn null_terminator_index(&self) -> usize;
+}
+impl NullTerminatorIndex for &[u8] {
+    fn null_terminator_index(&self) -> usize {
+        let mut len = self.len();
+        {
+            while len > 0 && self[len - 1] == 0 {
```
Comment (wimax-grapl) [inline src/cpuid/src/brand_string.rs:303]: The major change here is moving the `len > 0` check before the `[]` accessor.

Ideally I'd do this as an .iter().rev().position() check but I assume there's a reason it's coded C style :)

File: src/cpuid/src/brand_string.rs [hunk 2]
```diff
@@ -118,14 +118,7 @@ impl BrandString {
...
 
-        let mut len = Self::MAX_LEN;
-        {
-            let this_bytes = this.as_bytes();
-            while this_bytes[len - 1] == 0 && len > 0 {
-                len -= 1;
-            }
-        }
-        this.len = len;
+        this.len = this.as_bytes().null_terminator_index();
```
Comment (serban300) [inline src/cpuid/src/brand_string.rs:121]: Could we use `this.as_bytes().trim_end_with().len()` ?

Comment (serban300) [reply src/cpuid/src/brand_string.rs:121]: Sorry, I didn't realize this.

File: src/cpuid/src/brand_string.rs [hunk 3]
```diff
@@ -298,13 +298,10 @@ trait NullTerminatorIndex {
...
 impl NullTerminatorIndex for &[u8] {
     fn null_terminator_index(&self) -> usize {
-        let mut len = self.len();
-        {
-            while len > 0 && self[len - 1] == 0 {
-                len -= 1;
-            }
+        match self.iter().rposition(|&b| b != b'\0') {
+            Some(idx) => idx + 1,
+            None => 0,
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/cpuid/src/brand_string.rs:303]: The problem is that `from_host_cpuid` is seemingly not unit-testable against arbitrary data, just real honest-to-goodness memory addresses that are different per-CPU :/

File: src/cpuid/src/brand_string.rs (hunk 2 above)
Comment (dianpopa) [reply src/cpuid/src/brand_string.rs:121]: We could use `iter().reposition()`. Something along the lines:
```
this.len = {
            if let Some(index) = this.as_bytes().iter().rposition(|&x| x != 0) {
                index + 1
            } else {
                0
            }
        };
```
WDYT?

Comment (dianpopa) [reply src/cpuid/src/brand_string.rs:121]: Oh, sorry I see that you are doing exactly that in your next commit. Cool
Make sure you squash the changes into one commit before merging.
Also use the 50/72 format for the commit message in order for the tests to pass.

File: src/cpuid/src/brand_string.rs [hunk 4]
```diff
@@ -300,6 +293,13 @@ impl BrandString {
     }
 }
 
+fn null_terminator_index(slice: &[u8]) -> usize {
```
Comment (serban300) [inline src/cpuid/src/brand_string.rs:296]: Nit: Another option would be to call this method `strlen` since I think it's equivalent to `strlen` from C.

Comment (wimax-grapl) [reply src/cpuid/src/brand_string.rs:296]: I'm a fan of overly-descriptive function names, but I can leave a comment about this if that's a fair compromise?
i'd argue that naming it strlen could have people unfamiliar with libc thinking they could swap this out with str::len

File: src/cpuid/src/brand_string.rs [hunk 5]
```diff
@@ -300,6 +293,13 @@ impl BrandString {
...
+    match slice.iter().rposition(|&b| b != b'\0') {
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/cpuid/src/brand_string.rs:297]: good point

File: src/cpuid/src/brand_string.rs (hunk 4 above)
Comment (serban300) [reply src/cpuid/src/brand_string.rs:297]: Agree. I think a comment describing more precisely what the function is doing would help.

**Comment by wimax-grapl** [comment]
(Also, just to understand Firecracker's flow, I've just been `git commit --amend`ing due to the above comment:
> Make sure you squash the changes into one commit before merging.
Do you just not do squashes on the Github PR page, instead preferring to do it manually once it's approved?)

**Comment by wimax-grapl** [comment]
Okay, commit message is below 50 (though Github seems to be changing my `#2914` into a `firecracker-microvm#2914`) on display - not at `git log` though.

This previously was failing some integration test about snapshots on 4.x - let's see if that still fails - otherwise, I think I'm good to go?

**Review by serban300 — APPROVED** [review]

**Comment by serban300** [comment]
Yes, looks good ! The tests have passed also.

**Review by georgepisaltu — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**